src/Forms/GridField/GridFieldVersionedState.php

```diff
@@ -0,0 +1,111 @@
+<?php
+namespace SilverStripe\Forms\GridField;
+
+use SilverStripe\Forms\GridField\GridField;
+use SilverStripe\Forms\GridField\GridField_ColumnProvider;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\Versioned\Versioned;
+use SilverStripe\Core\Convert;
+
+class GridFieldVersionedState implements GridField_ColumnProvider
+{
+    protected $column = null;
+
+    protected $versionedLabelFields = ['Name', 'Title'];
+
+    /**
+     * Modify the list of columns displayed in the table.
+     *
+     * @see {@link GridFieldDataColumns->getDisplayFields()}
+     * @see {@link GridFieldDataColumns}.
+     *
+     * @param GridField $gridField
+     * @param array $columns List reference of all column names.
+     */
+    public function augmentColumns($gridField, &$columns)
+    {
+        $model = $gridField->getModelClass();
+        $isModelVersioned = $model::has_extension(Versioned::class);
+
+        if(!$isModelVersioned) {
+            return;
+        }
+
+        $matchedVersionedFields = array_intersect(
+            $columns,
+            $this->versionedLabelFields
+        );
+
+        if (count($matchedVersionedFields) > 0) {
+            $this->column = array_values($matchedVersionedFields)[0];
+        }
+        // Use first column
+        else if ($columns) {
+            $this->column = $columns[0];
+        }
+    }
+
+    /**
+     * Names of all columns which are affected by this component.
+     *
+     * @param GridField $gridField
+     * @return array
+     */
+    public function getColumnsHandled($gridField)
+    {
+        return [$this->column];
+    }
+
+    /**
+     * HTML for the column, content of the <td> element.
+     *
+     * @param  GridField $gridField
+     * @param  DataObject $record - Record displayed in this row
+     * @param  string $columnName
+     * @return string - HTML for the column. Return NULL to skip.
+     */
+    public function getColumnContent($gridField, $record, $columnName)
+    {
+
+        $flagContent = '';
+        $flags = $record->getStatusFlags();
```

Comment:
From memory `getStatusFlags()` is specific to `SiteTree` and other Versioned `DataObject` do not have this method.
The `getStatusFlags()` defined in the `Versioned` extension doesn't return the same data signature either, so it's quite confusing to follow.
SiteTree returns:
```
$status = [
  'draft' => ['title' => 'Bob', 'text' => 'Bob Text'],
];
```
Versioned returns:
```
$status = ['draft'];
```

I suggest building a `getStatusFlag()` protected method in this class and standardise the data which this `getColumnContent()` method could use rather than jump around so many variables which `SiteTree` introduces...
Focus on only showing `draft` and `modified` statuses, and maybe have `archived`/`removedfromdraft` as a nice-to-have.
I should be able to replace `SiteTree` with another Versioned `DataObject` like `File` or (custom) `MyVersionedProduct`. This is important for contentblocks/elemental, for example.